content/named-tuples.md

```diff
@@ -0,0 +1,777 @@
+```
+We would need an explicit `_.toTuple` selection to express this:
+```scala
+  val pairs: List[(String, Int)] = persons.map(_.toTuple)
+```
+Note that conformance rules for named tuples are analogous to the rules for named parameters. One can assign parameters by position to a named parameter list.
+```scala
+  def f(param: Int) = ...
+  f(param = 1)   // OK
+  f(2)           // Also OK
+```
+But one cannot use a name to pass an argument to an unnamed parameter:
+```scala
+    val f: Int => T
+    f(2)         // OK
+    f(param = 2) // Not OK
+```
+The rules for tuples are analogous. Unnamed tuples conform to named tuple types, but the opposite requires a conversion.
+
+### Pattern Matching
+
+When pattern matching on a named tuple, the pattern may be named or unnamed.
+If the pattern is named it needs to mention only a subset of the tuple names, and these names can come in any order. So the following are all OK:
+```scala
+Bob match
+  case (name, age) => ...
+
+Bob match
+  case (name = x, age = y) => ...
+
+Bob match
+  case (age = x) => ...
+
+Bob match
+  case (age = x, name = y) => ...
+```
+
+### Expansion
+
+Named tuples are in essence just a convenient syntax for regular tuples. In the internal representation, a named tuple type is represented at compile time as a pair of two tuples. One tuple contains the names as literal constant string types, the other contains the element types. The runtime representation of a named tuples consists of just the element values, whereas the names are forgotten. This is achieved  by declaring `NamedTuple`
+in package `scala` as an opaque type as follows:
+```scala
+  opaque type NamedTuple[N <: Tuple, +V <: Tuple] >: V = V
+```
+For instance, the `Person` type would be represented as the type
+```scala
+NamedTuple[("name", "age"), (String, Int)]
+```
+`NamedTuple` is an opaque type alias of its second, value parameter. The first parameter is a string constant type which determines the name of the element. Since the type is just an alias of its value part, names are erased at runtime, and named tuples and regular tuples have the same representation.
+
+A `NamedTuple[N, V]` type is publicly known to be a supertype (but not a subtype) of its value paramater `V`, which means that regular tuples can be assigned to named tuples but not _vice versa_.
+
+The `NamedTuple` object contains a number of extension methods for named tuples hat mirror the same functions in `Tuple`. Examples are
+`apply`, `head`, `tail`, `take`, `drop`, `++`, `map`, or `zip`.
+Similar to `Tuple`, the `NamedTuple` object also contains types such as `Elem`, `Head`, `Concat`
+that describe the results of these extension methods.
+
+The translation of named tuples to instances of `NamedTuple` is fixed by the specification and therefore known to the programmer. This means that:
+
+ - All tuple operations also work with named tuples "out of the box".
+ - Macro libraries can rely on this expansion.
+
+### Computed Field Names
```

Comment:
I like this feature since it makes Selectable a lot nicer to use. Combined with `NamedTuple.From`, it makes it possible to create powerful proxy types like the `Expr` example in the appendix.

However, there's something a bit odd about using NamedTuple (where the order of the elements matters) to define the members of a Selectable (where there is no ordering). This is fine as long as we can construct a NamedTuple that contains the members we'd like to select on, but if we rely on `NamedTuple.From` we're limited to the parameters of case classes and tuples because they have a notion of ordering but other types don't.

To lift this limitation, I think it'd be useful to define another magic operator to construct a NamedTuple from the public fields of a class or trait, the elements of the NamedTuple should be sorted alphabetically so that reordering `vals` in a class would keep being a source and binary compatible change:
```scala
trait Base:
  val y: Int = 3
class Foo extends Base:
  val z: Int = 1
  val x: Int = 2

NamedTuple.FieldsOf[Foo] =:= (x: Int, y: Int, z: Int)
```

Reply:
>To lift this limitation, I think it'd be useful to define another magic operator to construct a NamedTuple from the public fields of a class or trait

If we add this, I don't see why `NamedTuple.From` cannot be used for both. Or at least a more clear name that will highlight the difference (`FieldsOf` does not).